theano/misc/pkl_utils.py

```diff
@@ -93,3 +113,217 @@ class CompatUnpickler(pickle._Unpickler):
+        name, in which case they will be numbered e.g. `x`, `x_2`, `x_3`, etc.
+        Defaults to ``True``.
+    :type allow_duplicates: bool, optional
+
+    :raises ValueError
+        If an unnamed shared variable is encountered and `allow_unnamed` is
+        ``False``, or if two shared variables have the same name, and
+        `allow_duplicates` is ``False``.
+
+    """
+    def __init__(self, zip_file, allow_unnamed=True, allow_duplicates=True):
+        super(PersistentSharedVariableID, self).__init__(zip_file)
+        self.name_counter = defaultdict(int)
+        self.ndarray_names = {}
+        self.allow_unnamed = allow_unnamed
+        self.allow_duplicates = allow_duplicates
+
+    def _resolve_name(self, obj):
+        if id(obj) in self.ndarray_names:
+            name = self.ndarray_names[id(obj)]
+            count = self.name_counter[name]
+            if count:
+                if not self.allow_duplicates:
+                    raise ValueError("multiple shared variables with the name "
+                                     "`{}` found".format(name))
+                name = '{}_{}'.format(name, count + 1)
+            self.name_counter[name] += 1
+            return name
+        return super(PersistentSharedVariableID, self)._resolve_name(obj)
+
+    def __call__(self, obj):
+        if isinstance(obj, SharedVariable):
+            if obj.name:
+                if obj.name == 'pkl':
+                    ValueError("can't pickle shared variable with name `pkl`")
+                self.ndarray_names[id(obj.container.storage[0])] = obj.name
+            elif not self.allow_unnamed:
+                raise ValueError("unnamed shared variable, {}".format(obj))
+        return super(PersistentSharedVariableID, self).__call__(obj)
+
+
+class PersistentNdarrayLoad(object):
+    """Load NumPy arrays that were persisted to a zip file when pickling.
+
+    :param zip_file: The zip file handle in which the NumPy arrays are saved.
+    :type zip_file: :class:`zipfile.ZipFile`
+
+    """
+    def __init__(self, zip_file):
+        self.zip_file = zip_file
+
+    def __call__(self, persid):
+        array_type, name = persid.split('.')
+
+        array = numpy.lib.format.read_array(self.zip_file.open(name))
+        if array_type == 'cuda_ndarray':
+            if config.experimental.unpickle_gpu_on_cpu:
+                # directly return numpy array
+                warnings.warn("config.experimental.unpickle_gpu_on_cpu is set "
+                              "to True. Unpickling CudaNdarray as "
+                              "numpy.ndarray")
+                return array
+            elif cuda_ndarray:
+                return cuda_ndarray.CudaNdarray(array)
+            else:
+                raise ImportError("Cuda not found. Cannot unpickle "
+                                  "CudaNdarray")
+        else:
+            return array
+
+
+def dump(obj, f, protocol=DEFAULT_PROTOCOL,
```

Comment:
The intent isn't really to introduce any sort of new behaviour. The way `CudaNdarray` variables are dealt with is just an exact copy of the `CudaNdarray_unpickler` function in `sandbox.cuda.type`. Crashing when unpickling without GPU is also fine by me (because the user can still read the parameters in the zip-file.) I'd prefer it though if we don't have to wait till the new GPU back-end.